cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java

```diff
@@ -37,7 +53,7 @@
  * It provides definition for {@link ServiceDiscoverer#getServiceURL}  and expects the sub-classes to give definition
  * for {@link AbstractServiceDiscoverer#getDiscoveryServiceClient}.
  */
-public abstract class AbstractServiceDiscoverer implements ServiceDiscoverer {
+public abstract class AbstractServiceDiscoverer implements ServiceDiscoverer, StreamContext {
```

Comment:
Actually... this gives the implementation to *all* program contexts (eg. `BasicFlowletContext`) but doesn't expose them through the interface. So technically I could do: `((StreamContext) getContext()).write("stream", "stream");` inside a flowlet process method and it would work. Do we want to allow that? If yes, then all the program contexts should extend `StreamContext`. If not, then the implementation should move into its own abstract class.

(Btw, why are worker's special when it comes to writing to streams? If we're exposing this capability shouldn't we expose it uniformly for all programs?)